sycl/doc/KernelProgramCache.md

```diff
@@ -68,42 +68,87 @@ times, which may involve JIT compilation and take quite a lot of time.
+interop kernels and programs.
+
+Also JIT compilation for cases when an application contains huge amount of
+device code (big kernels or multiple kernels) may take significant time. The
+kernels and programs are rebuild on every program restart. AOT compilation
+can be used to avoid that but it ties application to specific backend runtime
+version(s) and predefined HW configuration(s). As general solution it is
+reasonable to have kernel/program cache which works between application
+restarts (e.g. cache on disk for devices code build for specific HW/SW
+configuration).
 
 <a name="what-is-program">1</a>: Here "program" means an internal SYCL runtime
 object corresponding to a SPIRV module or native binary defining a set of SYCL
 kernels and/or device functions.
 
 
 ## Data structure of cache
+The cache is split into two levels:
+ - in-memory cache which is used during application runtime for device code
+ which has been already loaded and built for target device.
+ - on-disk cache which is used to store device binaries between application
+ executions.
 
+### In-memory cache
 The cache stores underlying PI objects behind `cl::sycl::program` and
 `cl::sycl::kernel` user-level objects in a per-context data storage. The storage
 consists of two maps: one is for programs and the other is for kernels.
 
-The programs map's key consists of three components: kernel set id<sup>[1](#what-is-ksid)</sup>,
-specialized constants, device this program is built for.
+The programs map's key consists of four components: kernel set id<sup>[1](#what-is-ksid)</sup>,
+specialized constants, device this program is built for, build options id <sup>[2](#what-is-bopts)</sup>.
 
-The krnels map's key consists of three components too: program the kernel
-belongs to, kernel name<sup>[2](#what-is-kname)</sup>, device the program is
-built for.
+The kernels map's key consists of four components too: program the kernel
+belongs to, kernel name<sup>[3](#what-is-kname)</sup>, device the program is
+built for, build options id<sup>[2](#what-is-bopts)</sup>.
 
 <a name="what-is-ksid">1</a>: Kernel set id is an ordinal number of the device
 binary image the kernel is contained in.
-
-<a name="what-is-kname">2</a>: Kernel name is a kernel ID mangled class' name
+<a name="what-is-bopts">2</a>: Hash for the string representation of build
+options set in application or environment variables (e.g.
+SYCL_PROGRAM_COMPILE_OPTIONS, SYCL_PROGRAM_LINK_OPTIONS). The key is hash out
+of the string which is concatenation of `program.get_compile_options()`,
+`program.get_link_options()`, `program.get_build_options()` values.
+<a name="what-is-kname">3</a>: Kernel name is a kernel ID mangled class' name
 which is provided to methods of `cl::sycl::handler` (e.g. `parallel_for` or
 `single_task`).
 
+### On-disk cache
+The cache is hidden behind in-memory cache and stores the same underlying PI
+object behind `cl::sycl::program` user-level objects in a per-context data
```

Comment:
General comment: `sycl::program` is being deprecated, and `sycl::module` will replace it, which has different API. This design should be flexible enough to easily migrate to modules in future. Otherwise it might make sense to implement modules first then the cache.

Reply:
No difficulties are seen in migration from sycl::program to sycl::kernel_bundle because cache is operating with underlining interfaces which are not going to be changed.